File: Exercise2.go
package main

import (
	"fmt"
)

type StackNode struct {
	Val  int
	Next *StackNode
}

type Stack struct {
	head *StackNode
}

func NewStack() *Stack {
	return &Stack{
		head: nil,
	}
}

func (s *Stack) isEmpty() bool {
	return s.head == nil
}

func (s *Stack) Push(x int) {
	newHead := &StackNode{Val: x}
	if s.head == nil {
		s.head = newHead
		return
	}
	newHead.Next = s.head
	s.head = newHead
}

func (s *Stack) Pop() int {
	if s.isEmpty() {
		panic("Stack is empty, nothing to pop....")
	}
	out := s.head.Val
	newHead := s.head.Next
	s.head.Next = nil
	s.head = newHead
	return out
}

func (s *Stack) Peek() int {
	if s.isEmpty() {
		panic("Stack is empty, nothing to pop....")
	}
	return s.head.Val
}

func (s *Stack) printStack() {
	if s.head == nil {
		return
	}
	msg := ""
	current := s.head
	for current != nil {
		msg += fmt.Sprintf("%v", current.Val)
		if current.Next != nil {
			msg += "->"
		}
		current = current.Next
	}
	fmt.Println(msg)
}

func main() {
	s := NewStack()
	s.Push(10)
	s.Push(20)
	s.Push(30)

	s.printStack()
	fmt.Println("Peek: ", s.Peek())
	s.printStack()

	top := s.Pop()
	fmt.Println("Top: ", top)
	s.printStack()
}

File: Exercise_2.swift
import Cocoa


// Time Complexity - O(n)
// Space Complexity - O(n)

class StackNode<T> {
    let value: T
    var next: StackNode<T>?
    
    init(value: T, next: StackNode<T>? = nil) {
        self.value = value
        self.next = next
    }
}

class Stack<T> {
    var top: StackNode<T>?
    
    var isEmpty: Bool {
        return top == nil
    }
    
    func push(newValue: T) {
        let newStackNode = StackNode(value: newValue)
        
        newStackNode.next = top
        top = newStackNode
    }
    
    func pop() -> T? {
        
        guard let topNode = top else {
            return nil
        }
        
        let newReturnNode = topNode
        self.top = topNode.next
        return newReturnNode.value
    }
    
    func peek() -> T? {
        return top?.value
    }
    
}

extension StackNode: CustomStringConvertible {
    var description: String {
        guard let next = next else {
            return "\(value)"
        }
        return "\(value) -> " + String(describing: next) + " "
    }
}

extension Stack: CustomStringConvertible {
    var description: String {
        guard let top = top else {
            return "Empty stack"
        }
        return String(describing: top)
    }
}

var stackObj = Stack<Int>()
print(stackObj.isEmpty)
print(stackObj.peek())
stackObj.push(newValue: 1)
stackObj.push(newValue: 2)
stackObj.push(newValue: 3)
stackObj.push(newValue: 4)
stackObj.push(newValue: 21)

print(stackObj.peek())

print(stackObj)

print(stackObj.pop())

print(stackObj)



